test/test_linalg.py

```diff
@@ -751,6 +751,155 @@ def test_det(self, device, dtype):
         with self.assertRaises(RuntimeError):
             op(t)
 
+    @skipCUDAIfNoMagma
+    @skipCPUIfNoLapack
+    @dtypes(torch.double, torch.cdouble)
+    # NOTE: This test, and many others in this file that use magma, are currently skipped for ROCm.
+    # See: https://github.com/pytorch/pytorch/issues/51303
+    @skipCUDAIfRocm
+    def test_det_autograd(self, device, dtype):
+        # Create a batch of randomized square singular matrices. Each matrix is
+        # generated by performing a matrix multiply of two random matrices of
+        # size (size[-1], 1) and (1, size[-1])
+        def make_singular_matrix_batch(size):
+            assert size[-1] == size[-2]
+
+            batch = torch.empty(size, device=device, dtype=dtype)
+
+            matrix_size = size[-2:]
+            view_size = ((1,) + size) if len(size) == 1 else ((int(np.prod(size[:-2])),) + matrix_size)
+            batch_view = batch.view(view_size)
+
+            a_size = (size[-1], 1)
+            b_size = (1, size[-1])
+
+            for matrix_idx in range(batch_view.size(0)):
+                # TODO: linalg_det_backward can sometimes produce NaNs if any
+                # singular values are 0. This is likely due to its dependence
+                # upon svd_backward. Therefore, we must avoid testing such
+                # matrices. Once lu_backward is available, linalg_det_backward
+                # can be updated to use it so that wo no longer have this
+                # problem: https://github.com/pytorch/pytorch/issues/53364
+                svd_has_zero = True
+
+                # Note: When calculating the determinant, floating point errors
+                # can be introduced such that the calculated determinant is
+                # a very small number, rather than exactly zero. We use random
+                # integers cast to float to drastically decrease the chance of
+                # this happening and then double check that the determinant is
+                # exactly zero. If determinant is nonzero, we regenerate
+                # a limited number of times until it is zero.
```

Comment:
We can probably avoid the generation process and rather rely on something deterministic as described in https://github.com/pytorch/pytorch/pull/54787/files#r608492429. We could use `torch.allclose` with proper `rtol, atol` to assess the closeness to zero.

Reply:
I see your point now, the code considers singular cases only if `det == 0`.